frame/bags-list/src/list/mod.rs

```diff
@@ -0,0 +1,670 @@
+		// as this bag is the correct one, we're good. All calls to this must come after getting the
+		// correct [`notional_bag_for`].
+		self.insert_node_unchecked(Node::<T> { id, prev: None, next: None, bag_upper: 0 });
+	}
+
+	/// Insert a node into this bag.
+	///
+	/// This is private on purpose because it's naive; it doesn't check whether this is the
+	/// appropriate bag for this node at all. Generally, use [`List::insert`] instead.
+	///
+	/// Storage note: this modifies storage, but only for the node. You still need to call
+	/// `self.put()` after use.
+	fn insert_node_unchecked(&mut self, mut node: Node<T>) {
+		if let Some(tail) = &self.tail {
+			if *tail == node.id {
+				// this should never happen, but this check prevents one path to a worst case
+				// infinite loop.
+				debug_assert!(false, "system logic error: inserting a node who has the id of tail");
+				crate::log!(warn, "system logic error: inserting a node who has the id of tail");
+				return
+			};
+		}
+
+		// re-set the `bag_upper`. Regardless of whatever the node had previously, now it is going
+		// to be `self.bag_upper`.
+		node.bag_upper = self.bag_upper;
+
+		let id = node.id.clone();
+		// update this node now, treating it as the new tail.
+		node.prev = self.tail.clone();
+		node.next = None;
+		node.put();
+
+		// update the previous tail.
+		if let Some(mut old_tail) = self.tail() {
+			old_tail.next = Some(id.clone());
+			old_tail.put();
+		}
+		self.tail = Some(id.clone());
+
+		// ensure head exist. This is only set when the length of the bag is just 1, i.e. if this is
+		// the first insertion into the bag. In this case, both head and tail should point to the
+		// same node.
+		if self.head.is_none() {
+			self.head = Some(id.clone());
+			debug_assert!(self.iter().count() == 1);
+		}
+	}
+
+	/// Remove a node from this bag.
+	///
+	/// This is private on purpose because it doesn't check whether this bag contains the node in
+	/// the first place. Generally, use [`List::remove`] instead, similar to `insert_unchecked`.
+	///
+	/// Storage note: this modifies storage, but only for adjacent nodes. You still need to call
+	/// `self.put()` and `ListNodes::remove(id)` to update storage for the bag and `node`.
+	fn remove_node_unchecked(&mut self, node: &Node<T>) {
+		// reassign neighboring nodes.
+		node.excise();
+
+		// clear the bag head/tail pointers as necessary.
+		if self.tail.as_ref() == Some(&node.id) {
+			self.tail = node.prev.clone();
+		}
+		if self.head.as_ref() == Some(&node.id) {
+			self.head = node.next.clone();
+		}
+	}
+
+	/// Sanity check this bag.
+	///
+	/// Should be called by the call-site, after any mutating operation on a bag. The call site of
+	/// this struct is always `List`.
+	///
+	/// * Ensures head has no prev.
+	/// * Ensures tail has no next.
+	/// * Ensures there are no loops, traversal from head to tail is correct.
+	fn sanity_check(&self) -> Result<(), &'static str> {
```

Comment:
should also call into the (to be made) `sanity_check` of all the nodes in the bag, and ensure that their `bag_upper` is this bag. 

Although, this is not complete. Imagine there's a stale node, as a result of a faulty migration, that is not part of any bag. This might not appear in any iterations and be skipped overall. I think next to this, we should call the `sanity_check` of all nodes in the `List::sanity_check` as well. A bit expensive, but probably worth it.

Reply:
as noted, I also want all of these functions to be `std` only. Then, we should add a lot of loggings to them, so that by calling `List::sanity_check`, not only you check that everything is absolutely correct, you also get a good insight into what are the stats and what was checked. Then, once we deploy this, we can call `List::sanity_check` hourly in a remote-externalities.

Reply:
b39e1c05e1f1a8cec49c5f04ca551c00aee0a713 -> adds a node sanity check and a little bit of extra logging within `List::sanity_check`